petstore/src/main/scala/io/finch/petstore/petstoreDb.scala

```diff
@@ -0,0 +1,322 @@
+        }
+      }
+    }
+
+  /**
+   * Helper method for addPet: Adds a [[Category]] to the category map.
+   * This differs from the Swagger example in that the Category's ID is autogenerated
+   * rather than passed by the user. This is done to avoid potential memory and storage
+   * sabotage. Category with given IDs will be rejected.
+   * @param c The Category we want to add
+   * @return The Category just added
+   */
+  private def addCat(c: Category): Future[Category] =
+    cat.synchronized {
+      c.id match{
+        case Some(x) => Future.exception(InvalidInput("New Category should not contain an id"))
+        case None => cat.synchronized{
+          val genId = if (cat.isEmpty) 0 else cat.keys.max + 1
+          cat(genId) = c.copy(id = Some(genId))
+          Future(cat(genId))
+        }
+      }
+    }
+
+  /**
+   * POST: Adds a [[Pet]] to the database, validating that the ID is empty.
+   *
+   * @param inputPet the new pet
+   * @return the id of the new pet
+   */
+  def addPet(inputPet: Pet): Future[Long] =
+    inputPet.id match {
+      case Some(_) => Future.exception(InvalidInput("New pet should not contain an id"))
+      case None => pets.synchronized {
+        val id = if (pets.isEmpty) 0 else pets.keys.max + 1
+        pets(id) = inputPet.copy(id = Some(id))
+
+        inputPet.tags match{
+          case Some(tagList) => tagList.map(addTag(_))
+          case None => None
+        }
+
+        inputPet.category match{
+          case Some(c) => addCat(c)
+          case None => None
+        }
+
+        Future.value(id)
+      }
+    }
+
+  /**
+   * PUT: Updates an existing [[Pet]], while validating that a current version of
+   * the [[Pet]] exists (a.k.a. an existing [[Pet]] has the same id as inputPet).
+   * @param inputPet The [[Pet]] we want to replace the current [[Pet]] with. Must be passed with the original Pet's ID.
+   * @return The updated pet
+   */
+  def updatePet(inputPet: Pet): Future[Pet] = inputPet.id match {
+    case Some(id) =>
+      if(pets.contains(id)) pets.synchronized{
+        pets(id) = inputPet
+        Future.value(inputPet)
+      } else {
+        Future.exception(MissingPet("Invalid id: doesn't exist"))
+      }
+    case None => Future.exception(MissingIdentifier(s"Missing id for pet: $inputPet"))
+  }
+
+  /**
+   * Helper method: Allows the user to get all the pets in the database.
+   * @return A sequence of all pets in the store.
+   */
+  private def allPets: Future[Seq[Pet]] = Future.value(
+    pets.synchronized(pets.toList.sortBy(_._1).map(_._2))
+  )
+
+  /**
+   * GET: Find pets by status
+   * @param s The status to filter Pets by.
+   * @return A sequence of all Pets with the given status.
+   */
+  def getPetsByStatus(s: Status): Future[Seq[Pet]] = {
+    val allMatchesFut = for{
+      petList <- allPets //List[Pet]
+      allBool = petList.map(_.status)
+    } yield petList.filter(_.status.map(_.code.equals(s.code)).getOrElse(false))
```

Comment:
`Option(foo).map(condition).getOrElse(false)` is better written as `Option(foo).exists(condition)`.